src/proj_4D_api.c

```diff
@@ -724,22 +724,74 @@ PJ_CONTEXT *proj_context_destroy (PJ_CONTEXT *ctx) {
 }
 
 
+
+
+
+
 /*****************************************************************************/
-PJ_INFO proj_info(void) {
+static char *path_append (char *buf, const char *app, size_t *buf_size) {
+/******************************************************************************
+    Helper for proj_info() below. Append app to buf, separated by a
+    semicolon. Also handle allocation of longer buffer if needed.
+
+    Returns buffer and adjusts *buf_size through provided pointer arg.
+******************************************************************************/
+    char *p;
+    size_t len, applen = 0, buflen = 0;
+
+    /* Nothing to do? */
+    if (0 == app)
+        return buf;
+    applen = strlen (app);
+    if (0 == applen)
+        return buf;
+
+    /* Start checking whether buf is long enough */
+    if (0 != buf)
+        buflen = strlen (buf);
+    len = buflen+applen+2;
+
+    /* "pj_realloc", so to speak */
```

Comment:
indeed, why not using realloc() ?

Reply:
For two reasons: First (and least), to get the null-padding from calloc.

Second: Because (for reasons beyond my comprehension) we use wrappers for memory allocation.

I would really like to get rid of pj_malloc, pj_calloc, pj_dalloc, and just use malloc, calloc and free. I would equally like to see the fileapi go, and just use direct stdio calls where needed. But I cannot convince myself that this will not break something for someone (in the case of removing fileapi, stuff will probably break for Google (@schwehr?), since I think Frank @warmerdam worked for Google when it was introduced.

For pj_malloc and friends: Yes - getting rid would be fantastic. Could be a thing to consider for v. 5.1.0 or 6.0.0?

Reply:
I think going with the standard malloc, calloc and free calls should be fine.  The one thing to watch out is that malicious input can't allocate huge amounts of memory.aq

The one thing that I need is a way to have the external data files not be required for a binary.  If there is some alternative way to do this, that would be great. 

My constraints, which are different from the normal linux install: containers have no local storage.  Each container runs one binary.  That binary can be "large."  There is a bit of local temp ramdisk, but it is so small that data files can't go there.

The same issue happens with GDAL and I solved that with a library that sadly isn't part of the public ABSL library release.  But then I stash them in /vsimem/ and set the GDAL_DATA env var.   In a perfect world, proj, libgeotiff, and gdal would all have the same way to stash support data in the binary that was an optional configure flag.

FYI, The other thing we do is hook in the google logger so that proj messages end up in the standard logs.

Reply:
> The same issue happens with GDAL and I solved that with a library that sadly isn't part of the public ABSL library release. But then I stash them in /vsimem/ and set the GDAL_DATA env var. In a perfect world, proj, libgeotiff, and gdal would all have the same way to stash support data in the binary that was an optional configure flag.

What sort of data are we talking about here? Grids and init-files?

> FYI, The other thing we do is hook in the google logger so that proj messages end up in the standard logs.

Are you needs sufficiently covered in the new API?

Reply:
The embedded data was indeed setup by Warmerdam.

This is indeed the init (epsg, proj_def.dat) and grid shift files.  BTW, looking in proj_def.dat, I was amused to see "SCCSID"

Logging is all set as long as there is pj_ctx_set_logger().